### PR TITLE
Hardware/test/new ipd buttons

### PR DESCRIPTION
**Pull request by bjornminderman**
Closes PM-255 Add rotary encoder to input device software

## Description

This pull request adds an interface in the software to readout and use the data from a rotary encoder to scroll through the IPD menu. This rotary encoder will replace the joystick on the IPD.  For the readout of the encoder an external open source library has been used.

Besides the encoder, also an additional button was added to create a  shortcut to the walking and single step gaits. As from now on two buttons make use of the double-push functionality, it was decided to make the double push button a separate class.

Please bare in mind that the switch to the rotary encoder, requires a new menu to be designed. At the moment only a simplified menu exists and the menu and certain functionalities will be expanded in following PRs.

## Changes

* Add rotary encoder library
* Use encoder output to scroll through menu
* Add a double click button class
* Add shortcut push button
* Create a simplified version of the new menu
* Remove joystick

### CODE REVIEW

File: include/state.h [hunk 1]
```diff
@@ -127,8 +113,8 @@ class State
 
   const State* left_ = this;
   const State* right_ = this;
-  const State* up_ = this;
-  const State* down_ = this;
+  const State* shortcutPush_ = this;
+  const State* shortcutDoublePush_ = this;
```
Comment (Olavhaasie) [inline include/state.h]: Member names should be underscore_case, so `shortcut_push_` and `shortcut_double_push_`

File: src/main.cpp [hunk 2]
```diff
@@ -148,7 +163,7 @@ void setup()
 
   // Reset the joystick right pin, this needed after the ROS node init pin 14 is
   // apparently used by ROS.
-  pinMode(pins::JOYSTICK_RIGHT, INPUT_PULLUP);
+  // pinMode(pins::JOYSTICK_RIGHT, INPUT_PULLUP);
```
Comment (Olavhaasie) [inline src/main.cpp]: Think this can be removed?

File: include/state_machine.h [hunk 3]
```diff
@@ -20,19 +20,20 @@ class StateMachine
...
 
 private:
-  void constructWalkMenu(State* from);
-  void constructSideStepMenu(State* from);
-  void constructStepMenu(State* from);
-  void constructSofaMenu(State* from);
-  void constructStairsMenu(State* from);
+  void constructObstacleMenu(State* from, State* walk, State* single_step, State* obstacles);
+  void constructWalkMenu(State* from, State* prev_gait);
+  // void constructSideStepMenu(State* from);
```
Comment (Olavhaasie) [inline include/state_machine.h]: Why has this been removed?

**Review by Olavhaasie — CHANGES_REQUESTED** [review]

File: include/state_machine.h (hunk 3 above)
Comment (bjornminderman) [reply include/state_machine.h]: Not exactly sure. But there is no side step menu so now I deleted it completely

File: src/main.cpp [hunk 4]
```diff
@@ -148,7 +163,6 @@ void setup()
 
   // Reset the joystick right pin, this needed after the ROS node init pin 14 is
   // apparently used by ROS.
```
Comment (Olavhaasie) [inline src/main.cpp]: This comment can also be removed

**Review by Olavhaasie — CHANGES_REQUESTED** [review]
Should be good

**Review by Olavhaasie — APPROVED** [review]

**Pull request merged**